File: kotlin/src/main/kotlin/domain/Control.kt
package domain

import shared.*
import java.time.Instant

/**
 * ControlStatus: sealed classによるADT
 * Kotlinのsealed classはコンパイル時に網羅性をチェック可能
 */
sealed class ControlStatus {
    data object NotImplemented : ControlStatus()

    data class InProgress(val progress: Percentage) : ControlStatus()

    data class Implemented(val implementedAt: Instant) : ControlStatus()

    data class NotApplicable(val reason: String) : ControlStatus()

    data class Failed(
        val reason: String,
        val detectedAt: Instant
    ) : ControlStatus()
}

/**
 * when式によるパターンマッチ
 * sealed classなのでelse不要（コンパイル時に網羅性保証）
 */
fun <T> ControlStatus.match(
    onNotImplemented: () -> T,
    onInProgress: (Percentage) -> T,
    onImplemented: (Instant) -> T,
    onNotApplicable: (String) -> T,
    onFailed: (String, Instant) -> T
): T = when (this) {
    is ControlStatus.NotImplemented -> onNotImplemented()
    is ControlStatus.InProgress -> onInProgress(progress)
    is ControlStatus.Implemented -> onImplemented(implementedAt)
    is ControlStatus.NotApplicable -> onNotApplicable(reason)
    is ControlStatus.Failed -> onFailed(reason, detectedAt)
}

/**
 * Control エンティティ
 * data classで不変性を保証（valのみ使用）
 */
data class Control private constructor(
    val id: ControlId,
    val frameworkId: FrameworkId,
    val code: String,
    val title: String,
    val description: String,
    val status: ControlStatus,
    val ownerId: UserId
) {
    companion object {
        fun create(
            id: String,
            frameworkId: FrameworkId,
            code: String,
            title: String,
            description: String,
            ownerId: UserId
        ): Result<Control, ValidationErrors> {
            val errors = mutableListOf<ValidationError>()

            val controlIdResult = ControlId.create(id)
            if (controlIdResult is Result.Err) {
                errors.add(controlIdResult.error)
            }

            if (code.isBlank()) {
                errors.add(ValidationError("code", "Control code is required", "REQUIRED"))
            }

            if (title.isBlank()) {
                errors.add(ValidationError("title", "Control title is required", "REQUIRED"))
            }

            if (errors.isNotEmpty()) {
                return err(ValidationErrors(errors))
            }

            return ok(Control(
                id = (controlIdResult as Result.Ok).value,
                frameworkId = frameworkId,
                code = code,
                title = title,
                description = description,
                status = ControlStatus.NotImplemented,
                ownerId = ownerId
            ))
        }
    }

    /**
     * copy()で不変性を保ちながら状態を更新
     * ビジネスルールのバリデーション付き
     */
    fun updateStatus(newStatus: ControlStatus): Result<Control, ValidationError> {
        // Business rule: Failed状態から直接Implementedには遷移できない
        if (status is ControlStatus.Failed && newStatus is ControlStatus.Implemented) {
            return err(ValidationError(
                "status",
                "Cannot transition directly from Failed to Implemented",
                "INVALID_TRANSITION"
            ))
        }

        return ok(copy(status = newStatus))
    }
}

/**
 * ステータスの表示用文字列
 */
fun ControlStatus.toLabel(): String = match(
    onNotImplemented = { "未実装" },
    onInProgress = { progress -> "実装中 (${progress.value}%)" },
    onImplemented = { implementedAt -> "実装済み ($implementedAt)" },
    onNotApplicable = { reason -> "適用外: $reason" },
    onFailed = { reason, _ -> "失敗: $reason" }
)

File: go/domain/control.go
// Package domain provides GRC domain models.
package domain

import (
	"fmt"
	"time"

	"github.com/example/grc-domain-models/domain/shared"
)

// ControlStatus represents the status of a control.
// Go doesn't have union types, so we use an interface with unexported method
// to create a "sealed" interface pattern.
type ControlStatus interface {
	controlStatus() // unexported method prevents external implementations
	String() string
}

// NotImplemented represents a control that hasn't been implemented.
type NotImplemented struct{}

func (NotImplemented) controlStatus() {}
func (NotImplemented) String() string { return "Not Implemented" }

// InProgress represents a control that is being implemented.
type InProgress struct {
	Progress shared.Percentage
}

func (InProgress) controlStatus() {}
func (s InProgress) String() string {
	return fmt.Sprintf("In Progress (%d%%)", s.Progress.Value())
}

// Implemented represents a control that has been implemented.
type Implemented struct {
	ImplementedAt time.Time
}

func (Implemented) controlStatus() {}
func (s Implemented) String() string {
	return fmt.Sprintf("Implemented (%s)", s.ImplementedAt.Format(time.RFC3339))
}

// NotApplicable represents a control that is not applicable.
type NotApplicable struct {
	Reason string
}

func (NotApplicable) controlStatus() {}
func (s NotApplicable) String() string {
	return fmt.Sprintf("Not Applicable: %s", s.Reason)
}

// Failed represents a control that has failed.
type Failed struct {
	Reason     string
	DetectedAt time.Time
}

func (Failed) controlStatus() {}
func (s Failed) String() string {
	return fmt.Sprintf("Failed: %s (detected at %s)", s.Reason, s.DetectedAt.Format(time.RFC3339))
}

// MatchControlStatus provides exhaustive pattern matching for ControlStatus.
// Note: Go cannot guarantee compile-time exhaustiveness.
func MatchControlStatus[T any](
	status ControlStatus,
	onNotImplemented func() T,
	onInProgress func(shared.Percentage) T,
	onImplemented func(time.Time) T,
	onNotApplicable func(string) T,
	onFailed func(string, time.Time) T,
) T {
	switch s := status.(type) {
	case NotImplemented:
		return onNotImplemented()
	case InProgress:
		return onInProgress(s.Progress)
	case Implemented:
		return onImplemented(s.ImplementedAt)
	case NotApplicable:
		return onNotApplicable(s.Reason)
	case Failed:
		return onFailed(s.Reason, s.DetectedAt)
	default:
		panic(fmt.Sprintf("unknown ControlStatus type: %T", status))
	}
}

// Control represents a compliance control entity.
// Fields are unexported to ensure immutability.
type Control struct {
	id          shared.ControlID
	frameworkID shared.FrameworkID
	code        string
	title       string
	description string
	status      ControlStatus
	ownerID     shared.UserID
}

// Getter methods for Control
func (c *Control) ID() shared.ControlID          { return c.id }
func (c *Control) FrameworkID() shared.FrameworkID { return c.frameworkID }
func (c *Control) Code() string                  { return c.code }
func (c *Control) Title() string                 { return c.title }
func (c *Control) Description() string           { return c.description }
func (c *Control) Status() ControlStatus         { return c.status }
func (c *Control) OwnerID() shared.UserID        { return c.ownerID }

// CreateControlInput holds the input for creating a Control.
type CreateControlInput struct {
	ID          string
	FrameworkID shared.FrameworkID
	Code        string
	Title       string
	Description string
	OwnerID     shared.UserID
}

// NewControl creates a new Control with validation.
func NewControl(input CreateControlInput) (*Control, error) {
	var errors shared.ValidationErrors

	id, err := shared.NewControlID(input.ID)
	if err != nil {
		if ve, ok := err.(shared.ValidationError); ok {
			errors = append(errors, ve)
		}
	}

	if input.Code == "" {
		errors.Add("code", "Control code is required", "REQUIRED")
	}

	if input.Title == "" {
		errors.Add("title", "Control title is required", "REQUIRED")
	}

	if errors.HasErrors() {
		return nil, errors
	}

	return &Control{
		id:          id,
		frameworkID: input.FrameworkID,
		code:        input.Code,
		title:       input.Title,
		description: input.Description,
		status:      NotImplemented{},
		ownerID:     input.OwnerID,
	}, nil
}

// WithStatus returns a new Control with the updated status.
// This preserves immutability by creating a new instance.
func (c *Control) WithStatus(newStatus ControlStatus) (*Control, error) {
	// Business rule: Cannot transition directly from Failed to Implemented
	if _, isFailed := c.status.(Failed); isFailed {
		if _, isImplemented := newStatus.(Implemented); isImplemented {
			return nil, shared.NewValidationError(
				"status",
				"Cannot transition directly from Failed to Implemented",
				"INVALID_TRANSITION",
			)
		}
	}

	return &Control{
		id:          c.id,
		frameworkID: c.frameworkID,
		code:        c.code,
		title:       c.title,
		description: c.description,
		status:      newStatus,
		ownerID:     c.ownerID,
	}, nil
}

// GetControlStatusLabel returns a localized label for the control status.
func GetControlStatusLabel(status ControlStatus) string {
	return MatchControlStatus(
		status,
		func() string { return "未実装" },
		func(p shared.Percentage) string { return fmt.Sprintf("実装中 (%d%%)", p.Value()) },
		func(t time.Time) string { return fmt.Sprintf("実装済み (%s)", t.Format(time.RFC3339)) },
		func(reason string) string { return fmt.Sprintf("適用外: %s", reason) },
		func(reason string, _ time.Time) string { return fmt.Sprintf("失敗: %s", reason) },
	)
}
